### ISSUE TITLE
split config.js into several files OR maintain translations separately from the settings

### ISSUE THREAD

**Issue by intradesign** [kiprotect/klaro#254]
Is it possible (and if so: how) to split the parameters of config.js into several parts (files)? I would like to separate the translations: {...} section from the other contents of the file, so that the translations can be maintained in a central file, which can be reused in every project and easily overwritten after an update, and the actual configuration (behaviour and apps).

And: could the hierarchy of the translations be redesigned for a better overview, for example

```
translations: {
    consentModal: {
        title: {
            en: 'A title in German',
            en: 'An English title',
            },
        description: {
            en: 'A description in German',
            en: 'An English description',
            },
        },
    consentNotice: {
        description: {
            de: '...',
            en: '...',
        },
[...]
```

And: Could parameters from the translations (central file) be overwritten or added in the local config file if necessary?

**Comment by adewes** [comment]
Hi @intradesign! You can overwrite the default  translation settings by specifying new values in your own config: Just define a `translations` entry in the config and add the entries that you'd like to overwrite. We don't want to split the translations into a separate file by default as we don't want to force people to include these translations separately, we could add an option to not load the default translations though so that you can import your own version.

**Comment by intradesign** [comment]
Hello Andreas,
thanks for your quick reply!

The config.js is exactly the place where I would like to get rid of the
translations and swap them to an external file instead. At the moment I
have all translations in the config.js.

My goal would be to separate the texts (which are identical for many
installations and only differ for a few) from the default settings and the
apps, which differ for each installation.

With default settings I mean: elementID, storageMethod, cookieName
CookieExpiresAfterDays, privacyPolicy, imprint, default, mustConsent,
acceptAll, hideDicliceAll, hideLearnMore

So the file with the translations can easily be maintained independently
and, if necessary, simply overwritten in different installations when it is
updated, without the risk of accidentally overwriting the default settings.
The translation file could then easily contain translation text for apps
that are not used in the respective instance.

To do this, of course, you would have to standardize the naming under Apps
for all instances.

In other words: Where can I maintain the standard translations if not in
the config.js? In the klaro.js? It's quite confusing because it's big. Or
have I missed something?

Thank you very much and best regards
Timo Mackowiak


Am Mi., 22. Juli 2020 um 15:19 Uhr schrieb Andreas Dewes <
notifications@github.com>:

**Comment by adewes** [comment]
Hey @intradesign, we now include a version of Klaro without the default translations, which makes it possible to maintain translations separately. There's an example here:

https://github.com/kiprotect/klaro/blob/master/dist/examples/custom-translations.html

In general Klaro will load its default translations (if they are defined) and then overwrite them with the translations defined in your config.js. Not sure if that's what you want, you can easily maintain a common translations file and use the newly exposed `updateConfig(configA, configB)` function to merge your translations and config settings. `klaro.updateConfig(configA, configB)` modifies `configA` in place and recursively copies over every key/value pair present in `configB` to `configA`. By default, values from `configB` will overwrite values in `configA`, you can pass `false` as a third argument to not overwrite anything though. So you could do e.g.

```js

// your app-specific config (with some custom translations and app settings)
const appConfig = {...}
// your default translations
const defaultTranslations = {...}
// your default config
const defaultConfig = {...}

klaro.updateConfig(appConfig, defaultTranslations, false) // fill in the default translations
klaro.updateConfig(appConfig, defaultConfig) // fill in the default config settings
```
Does that solve your problem? Changes are in master but not yet deployed btw.

**Comment by intradesign** [comment]
Hi @adewes, that sounds really good. I will test it in the next few days.

**Comment by intradesign** [comment]
Hi adewes,

If I have understood this correctly, the standard translations have been removed from klaro.js and this file is now divided into two files klaro-no-translations.js and translations.js.

In the file translations.js I can maintain my very own standard texts and translations to be used for several projects.

Different translations and texts can still be maintained in the file config.js.

But I did not understand what I need the following functions for now:
```
klaro.updateConfig(appConfig, defaultTranslations, false) // fill in the default translations
klaro.updateConfig(appConfig, defaultConfig, false) // fill in the default config settings
```

and what that exactly means:
```
// your app-specific config (with some custom translations and app settings)
const appConfig = {...}
// your default translations
const defaultTranslations = {...}
// your default config
const defaultConfig = {...}
```

I hope you can help me with that.

Thanks and best regards,
Timo

**Comment by adewes** [comment]
Hi @intradesign, sorry for the late answer! The `klaro.updateConfig(a, b)` recursively updates object `a` with entries from object `b`. In the context of Klaro, `b` could e.g. contain translations or app configs, and `a` could be a base config that you want to use e.g. across many sites. `updateConfig` thereby enables you to construct configs from different snippets.

The `appConfig` would be the final config you use for your local Klaro installation. The `defaultTranslations` would be translations you want to add to your config, the `defaultConfig` would e.g. contain apps that you want to reuse across many websites.

`klaro.js` still contains the standard translations like before, but we now have a version `klaro-no-translations.js` that doesn't contain the translations, which makes more sense for cases where you e.g. want to load translations dynamically based on page language in order to save some bandwidth (though the translation files are quite small with around 1.5 kB each).

**Comment by adewes** [comment]
Hey @intradesign, closing this here for now, if you have any more questions just comment here or open another issue as usual :)
